portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java

```diff
@@ -0,0 +1,112 @@
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.DatastoreServiceConfig;
+import com.google.appengine.api.datastore.DatastoreServiceConfig.Builder;  
+import com.google.appengine.api.datastore.ReadPolicy;
+import com.google.appengine.api.datastore.ReadPolicy.Consistency;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/nickname")
+public class NicknameServlet extends HttpServlet {
+
+  /**
+   * Handles GET request. If the user is logged in, displays a page for them to input 
+   * nickname. This servlet will only be sent a request to if the logged in user 
+   * does not yet have a nickname.
+   */ 
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+    PrintWriter out = response.getWriter();
+    out.println("<h1>Set Nickname</h1>");
+
+    UserService userService = UserServiceFactory.getUserService();
+    String nickname = getUserNickname(userService.getCurrentUser().getUserId());
+    out.println("<p>Set your nickname here:</p>");
+    out.println("<form method=\"POST\" action=\"/nickname\">");
+    out.println("<input name=\"nickname\" value=\"" + nickname + "\" />");
+    out.println("<br/>");
+    out.println("<button>Submit</button>");
+    out.println("</form>");
+  }
+
+  /**
+   * Handles POST request by putting a nickname into Datastore and 
+   * associate it with a user id. 
+   */ 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/nickname");
+      return;
+    }
+
+    String nickname = request.getParameter("nickname");
+    String userId = userService.getCurrentUser().getUserId();
+
+    DatastoreService datastore = getDatastoreServiceWithConsistency();
+    Entity entity = new Entity("UserInfo", userId);
+    entity.setProperty("userId", userId);
+    entity.setProperty("nickname", nickname);
+
+    datastore.put(entity);
+
+    response.sendRedirect("/index.html");
+  }
+
+  /**
+   * Returns the nickname of the user with id, or null if the user has not set a nickname.
+   */
+  private String getUserNickname(String userId) {
```

Comment:
This is used in a couple places so you could extract this as well out into a helper class so you don't have duplicate code in your servlets. Also the same comment about using `Optional`.

Reply:
Moved to helper class Util.